### ISSUE TITLE
"ng update" doesn't support private repos such as FontAwesome 5 Pro

### ISSUE THREAD

**Issue by bjornharvold** [angular/angular-cli#10624]
### Versions
```
Angular CLI: 6.0.0
Node: 9.11.1
OS: darwin x64
Angular: 6.0.0
... animations, cli, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.6.0
@angular-devkit/build-angular     0.6.0
@angular-devkit/build-optimizer   0.6.0
@angular-devkit/core              0.6.0
@angular-devkit/schematics        0.6.0
@ngtools/webpack                  6.0.0
@schematics/angular               0.6.0
@schematics/update                0.6.0
rxjs                              6.1.0
typescript                        2.7.2
webpack                           4.6.0
```

### Repro steps
* Add dependency on a private repo such as FontAwesome 5 Pro
* Run ng update @angular/core

### Observed behavior
```
Not found : @fortawesome/pro-regular-svg-icons
```

### Mention any other details that might be useful (optional)
This works fine using npm directly because the private repo has already been registered with npm

**Comment by Ben305** [comment]
This should be fixed sooner than later

**Comment by Timebutt** [comment]
There is definitely a problem when using a private npm repository. I could only run `ng update` correctly once I removed my `.npmrc` file, and removed every private package from `package.json`. After running `ng update @angular/cli'` it finally correctly generated the `angular.json` file and I readded all the references I removed before.

**Comment by clydin** [comment]
Can you provide the list of option names used within `.npmrc`?

**Comment by Timebutt** [comment]
Sure, this is what my `.npmrc` looks like:

```
registry=https://npm.showpad.io:8080
save-exact=true
//npm.showpad.io:8080/:_authToken=${NPM_TOKEN}
```

**Comment by DaSchTour** [comment]
While looking at the growing list of bugs relating `.npmrc` #10704 #10571 #10660 I think this should be fixed asap

**Comment by hansl** [comment]
Hi @bjornharvold, @DaSchTour,

Could you verify this is still an issue with 6.0.1?  There were a few fixes regarding the registry flag (we have 1 fix in queue for the strict-ssl flag). Thanks!

**Comment by dimkanewtown** [comment]
```
    _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 6.0.1
Node: 9.2.1
OS: win32 x64
Angular: 6.0.0
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.6.1
@angular-devkit/build-angular     0.6.1
@angular-devkit/build-optimizer   0.6.1
@angular-devkit/core              0.6.1
@angular-devkit/schematics        0.6.1
@angular/cdk                      6.0.1
@angular/cli                      6.0.1
@angular/material                 6.0.1
@ngtools/webpack                  6.0.1
@schematics/angular               0.6.1
@schematics/update                0.6.1
rxjs                              6.1.0
typescript                        2.7.2
webpack                           4.6.0
```
Yep still happens with 6.0.1. Targeting VSTS in my .npmrc registry=https://PRIVATE.pkgs.visualstudio.com/_packaging/PRIVATE/npm/registry/
```
Unexpected token T in JSON at position 0
TF400813: Resource not available for anonymous access. Client authentication required.
```

**Comment by ThYpHo0n** [comment]
`Angular CLI: 6.0.1
Node: 8.9.0
OS: darwin x64
Angular: 6.0.1
... animations, cli, common, compiler, compiler-cli, core, forms
... http, platform-browser, platform-browser-dynamic
... platform-server, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.6.1
@angular-devkit/build-angular     0.6.1
@angular-devkit/build-optimizer   0.6.1
@angular-devkit/core              0.6.1
@angular-devkit/schematics        0.6.1
@angular/cdk                      5.2.5
@angular/material                 5.2.5
@ngtools/json-schema              1.1.0
@ngtools/webpack                  6.0.1
@schematics/angular               0.6.1
@schematics/update                0.6.1
ng-packagr                        2.4.4
rxjs                              6.1.0
typescript                        2.7.2
webpack                           4.6.0`
Still valid for me too using a scoped repository.

**Comment by abbazabacto** [comment]
Experiencing this issue with `@angular/cli@6.0.3` and private Nexus repository

Content of `.npmrc` looks something like...
```
registry=
email=
always-auth=true
_auth=
```

**Comment by tilowestermann** [comment]
Having the same issue using a JFrog Artifactory (universal artifact manager) as registry in `.npmrc`.

**Comment by ajpierson** [comment]
I'm also having this same problem with @angular/cli@6.0.3 and VSTS

**Comment by pouyio** [comment]
Same problem here behind a private Nexus repo

**Comment by eitland** [comment]
Same problem with 6.0.7 just now.

**Comment by tilowestermann** [comment]
@hansl any update on this? The issue still carries the "need: more info" label.

**Comment by Timebutt** [comment]
I'm also curious, is there anything we can do or provide to speed this issue up?

**Comment by devoto13** [comment]
I think it is fixed by https://github.com/angular/devkit/pull/982 in https://github.com/angular/devkit/releases/tag/v6.1.0-beta.1 release. Can anybody verify?

**Comment by ajpierson** [comment]
I'd be happy to test it, but I'm not quite sure how to...I'm a little confused how devkit relates to CLI.  What do I install to test this, is there an NPM package?  Do I clone the repo and do an NPM link? If I install that package does the CLI use it, or does it supersede the CLI.  Sorry, I'm a little ignorant, I'd love a little guidance.

**Comment by abbazabacto** [comment]
@devoto13 I can't seem to find `@angular/cli@6.1.0-beta.1`, it is not available in the npm registry: https://www.npmjs.com/package/@angular/cli

I was able to install `@angular-devkit/build-angular@0.7.0-beta.1` individually, but that still resulted in a `401 Unauthorized`, but not sure if this is related.

**Comment by devoto13** [comment]
@abbazabacto It is `@schematics/update@0.7.0-beta.1` package, which contains the fix. It is a dependency of `@angular/cli` and the last version depends on `@schematics/update@0.7.0-beta.0`. 

@ajpierson It looks like there is no easy way to try it out at the moment and the easiest way, as you suggest, is to wait until `@angular/cli@6.1.0-beta.1` is released. Sorry for the confusion.

**Comment by ajpierson** [comment]
Thanks, that's how it looked to me too as free I dug in some more glad to
hear I'm not crazy.

**Comment by joehansen** [comment]
Same Problem with `angular-cli 6.1.0-beta.2` :-(

**Comment by cluetjen** [comment]
We're using a private repository and get "error 401 unauthorized" (see #10704 which was closed as duplicate of this issue). The problem still exists with @angular/cli@6.1.0-rc.0.

**Comment by jppellerin** [comment]
For what it's worth, I was having the same problem and I simply removed all my private dependencies from `package.json`.  Then ran `ng update @angular/cli`.  Once that successfully completed, I re-added my dependencies to the `package.json` file.

**Comment by Brampage** [comment]
Issue still exists inside @angular/cli@6.0.8

**Comment by admosity** [comment]
I have the same issue with a private git+ssh dependency and trying this against the latest version on master:

```
Angular CLI: local (v6.1.0-beta.0, branch: master)
Node: 9.10.0
OS: darwin x64
Angular: 5.2.11
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router, service-worker

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.7.0-rc.0
@angular-devkit/build-angular     0.7.0-rc.0
@angular-devkit/build-optimizer   0.7.0-rc.0
@angular-devkit/build-webpack     0.7.0-rc.0
@angular-devkit/core              0.7.0-rc.0
@angular-devkit/schematics        0.7.0-rc.0
@angular/cdk                      5.2.5
@angular/cli                      1.7.0
@ngtools/webpack                  1.10.0
@schematics/angular               0.7.0-rc.0
@schematics/package-update        0.7.0-rc.0
@schematics/update                0.7.0-rc.0
rxjs                              5.5.11
typescript                        2.5.3
webpack                           4.9.2
```

**Comment by DavidDecraene** [comment]
Also having the not found issue using while using a git+https dependency running @angular/cli 6.0.8

Seems like a common bug, any timeline on the resolution of this?

**Comment by epelc** [comment]
Any chance the cli could just log not found errors as a warning instead of exiting on them?

**Comment by tengis** [comment]
Hi Guys,

The issue related with npm's get config command itself. 

`npm get <key>` returns all other configs except `_auth` and `_authToken`. Even, `npm config list` doesn't return them as well. 

On update script, it tries to run this command, and this command fails, then actual value became `null`, so your private registry authentication fails.

https://github.com/angular/angular-cli/blob/6449a753641340d8fc19a752e1a1ced75f974efa/packages/schematics/update/update/npm.ts#L47

So quick and dirty solution is set your auth token manually on **./node_modules/@schematics/update/update/npmjs** line number somewhere between 106 - 110.

`
auth = { token: "<your_auth_token", alwaysAuth: true }
`

FYI: My .npmrc file 
`
strict-ssl=false
registry=https://<company_nexus_npm_registry>
_auth="<base64_token>"
always_auth=true
email=<company_email>
`

I'll try to create PR for this if anyone created anything similar. 

Cheers

**Comment by Timebutt** [comment]
@tengis that's some great debugging friend! I just tried this out, and it works beautifully ;)

Important notice for people trying out this fix too, you need to have at least version `6.1.0-rc0` of the `@angular/cli` package, before editing the `npm.js` file as suggested.

What PR do you suggest to create, one that fixes the issue in `npm` itself, or one that 'patches' it in CLI by fetching it somehow from `.npmrc` and adding it in manually?

**Comment by Timebutt** [comment]
Thinking about it, you probably won't be able to change this in `npm` itself as `_auth` or `_authToken` are secrets, and it is desired behaviour to never hand them out using `npm get {KEY}`, judging from following terminal output:

```
npm get _authToken                                                                                                                                                         
npm ERR! ---sekretz---
```

**Comment by tengis** [comment]
@Timebutt Or perhaps add **--auth** flag on **ng update** command?

**Comment by DaSchTour** [comment]
I really wonder why not using plain npm and make use of the build in authentication. Adding an --auth flag so I have to add authentication again for angular. Very bad idea. Parse `.npmrc` or just use `npm`.

**Comment by jusefb** [comment]
We have the same issue now. Does anyone have a solution for this problem?

**Comment by beanian** [comment]
still seeing this behavior in @angular/cli@6.1.0-rc.2

**Comment by modul8com** [comment]
even in official 6.1 it's still an issue

**Comment by DaSchTour** [comment]
Well I don't think that this will be solved any time soon, just guessing by how it is labled.

**Comment by roblarsen** [comment]
**The good news**: I found a bug that precisely describes the issue we're experiencing. 
**The bad news**: It's been open since May 

😭

**Comment by tommck** [comment]
my hack/solution:

go to `node_modules\@schematics\update\update\npm.js` and change line 35-ish (first line of the method)

```js
function getNpmClientSslOptions(strictSsl, cafile) {
    const sslOptions = { strict: false }; // <---- RIGHT HERE.. normally, it's "{}"
    if (strictSsl === 'false') {
        sslOptions.strict = false;
    }
    else if (strictSsl === 'true') {
        sslOptions.strict = true;
    }
    if (cafile) {
        sslOptions.ca = fs_1.readFileSync(cafile);
    }
    return sslOptions;
}
```

Not sure why it's not getting the setting from the .npmrc file, but this let me get past the issue

**Comment by tilowestermann** [comment]
@hansl & @filipesilva This thread still has the "need more info" and "low frequency" labels. Based on the thread's activity, I assume that both are not valid anymore and possibly give a false impression of the severity of this issue.

**Comment by grantcp** [comment]
I'm also seeing this issue on various 6.0+ versions. Tried looking into the solutions above but didn't have any of the lines in my npm.js

Any timeline on a fix?

**Comment by e-hein** [comment]
still the same issue with @schematics/update@0.7.2 and @angular/cli@6.1.2

```
Angular CLI: 6.1.2
Node: 8.11.3
OS: darwin x64
Angular: 6.0.7
... common, compiler, compiler-cli, core, forms, http
... platform-browser, platform-browser-dynamic

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.6.8
@angular-devkit/build-angular     0.6.8
@angular-devkit/build-optimizer   0.6.8
@angular-devkit/core              0.0.29
@angular-devkit/schematics        0.0.52
@angular/animations               6.0.9
@angular/cli                      6.1.2
@angular/router                   6.0.9
@ngtools/json-schema              1.1.0
@ngtools/webpack                  6.0.8
@schematics/angular               0.7.2
@schematics/update                0.7.2
rxjs                              6.2.1
typescript                        2.7.2
webpack                           4.8.3
```
```
ng update @angular/cli
401 Unauthorized
```

Workaround that works for me:
-------------------------------
* temporary removing all private dependencies from package.json (no need to uninstall / reinstall them)
* and ensure there's no private default registry in .npmrc (usually I got the private mirror with public and private packages there)

**Comment by mrosshcm** [comment]
Still failing with `6.1.3`:

```
Angular CLI: 6.1.3
Node: 10.8.0
OS: linux x64
Angular: 6.1.2
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.7.3
@angular-devkit/build-angular     0.7.3
@angular-devkit/build-optimizer   0.7.3
@angular-devkit/build-webpack     0.7.3
@angular-devkit/core              0.7.3
@angular-devkit/schematics        0.7.3
@angular/cli                      6.1.3
@ngtools/webpack                  6.1.3
@schematics/angular               0.7.3
@schematics/update                0.7.3
rxjs                              6.2.2
typescript                        2.9.2
webpack                           4.9.2
```

```
$ ng update
401 Unauthorized
```

**Comment by DaSchTour** [comment]
Already page one of the most commented open issues. Hopefully we will get some attentation when this hits the 400 comments mark and reach the top of most commented issues. So come on people, we can do this.

**Comment by deniskrafczyk** [comment]
Facing this error too:

`
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 6.1.3
Node: 8.11.3
OS: win32 x64
Angular: 6.1.2
... animations, common, compiler, core, forms, http
... language-service, platform-browser, platform-browser-dynamic
... router

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.7.3 (cli-only)
@angular-devkit/core         0.7.3 (cli-only)
@angular-devkit/schematics   0.7.3 (cli-only)
@angular/cdk                 6.4.3
@angular/cli                 <error>
@angular/compiler-cli        <error>
@angular/flex-layout         6.0.0-beta.17
@schematics/angular          0.7.3 (cli-only)
@schematics/update           0.7.3 (cli-only)
rxjs                         6.2.2
typescript                   2.9.2 (cli-only)`

**Comment by aghou** [comment]
delete `node_modules` and run `ng upgrade @angular/core`
```shell
rm -rf node_modules
ng update @angular/core
```
it works for me

**Comment by mrosshcm** [comment]
@aghou That does provide a usable workaround, but the time spent removing and re-installing is not good.

Still hoping for a fix.

**Comment by Konni1337** [comment]
I have the same problem with 6.1.3 and JFrog Artifactory. A fix would be most appreciated.

.npmrc is like:
registry=
email=
always-auth=true
_auth=

**Comment by jziggas** [comment]
It's not just an issue with private repos, I get the same error with a public repo url that uses the "git+https://URL#HASH" pattern, e.g. :

`https://github.com/danhooper/ngQuickDate.git#e313316be410887d7289f355fc55ad4f4fecec15`

**Comment by mrmaxsteel** [comment]
Confirmation that @angular/cli 6.0.0+ has broken authentication to private npm registries:
```
$ ng update <package_on_private_registry>
Unable to authenticate, need: Basic realm="Artifactory Realm"
```

I have tested edge version @angular/cli@6.2.0-beta.3, and same issue. This is a serious bug that will prevent our organisation from upgrading to angular 6.

Please fix asap!

**Comment by akisiel1** [comment]
Same here. We have a private repo in VSTS and I cannot update `@angular/core` because of that bug. I vote for fixing it asap...

**Comment by blueiceprj** [comment]
Hi,

Globally set these values so it works in any project:
> npm config set "@fortawesome:registry" https://npm.fontawesome.com/ && \
   npm config set "//npm.fontawesome.com/:_authToken" XXXXXXX-XXXXXXXX-XXXXX

Its worked for me "ng update".

**Comment by dottodot** [comment]
The above works for me too.

**Comment by ankemp** [comment]
@blueiceprj solution might work for one off private repos, but not when your corp caches/replicates npm on jFrog Artifactory.

**Comment by daniel-leroux** [comment]
@ankemp Right! We are also blocked to perform the update to ^6.1.0

**Comment by williamzhao87** [comment]
Having same issue

**Comment by beanian** [comment]
+1 this really needs a resolution. Due to this issue we have missed our window to upgrade and have to wait till next release cycle in several months time 👎

**Comment by trojanc** [comment]
Really angular team...still no fix for this?! Stop developing any new stuff and fix this issues.
I have same issue, except I don't use private packages, nexus is used as a proxy for ALL npm repositories. So removing .npmrc doesn't solve anything (npm requests must go through the nexus proxy or they don't leave the network), and there is no "private dependencies" to remove, because they are all legit packages ...like angular stuff...

**Comment by smnbbrv** [comment]
Looks like I found a solution, at least for Nexus repository. I already filed this as a pull request, however it would be really good to have some feedback from you guys. Especially would be cool to have some Artifactory users' feedback.

How to try it right now: replace the `your_project/node_modules/@schematics/update/update/npm.js` with the following content

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const url = require("url");
const RegistryClient = require('npm-registry-client');
const npmPackageJsonCache = new Map();
const npmConfigOptionCache = new Map();
function _readNpmRc() {
    return new rxjs_1.Observable(subject => {
        // TODO: have a way to read options without using fs directly.
        const path = require('path');
        const fs = require('fs');
        const perProjectNpmrc = path.resolve('.npmrc');
        let npmrc = '';
        if (fs.existsSync(perProjectNpmrc)) {
            npmrc = fs.readFileSync(perProjectNpmrc).toString('utf-8');
        }
        else {
            if (process.platform === 'win32') {
                if (process.env.LOCALAPPDATA) {
                    npmrc = fs.readFileSync(path.join(process.env.LOCALAPPDATA, '.npmrc')).toString('utf-8');
                }
            }
            else {
                if (process.env.HOME) {
                    npmrc = fs.readFileSync(path.join(process.env.HOME, '.npmrc')).toString('utf-8');
                }
            }
        }
        const allOptionsArr = npmrc.split(/\r?\n/).map(x => x.trim());
        const allOptions = {};
        allOptionsArr.forEach(x => {
            const [key, ...value] = x.split('=');
            allOptions[key] = value.join('=');
        });
        subject.next(allOptions);
        subject.complete();
    }).pipe(operators_1.catchError(() => rxjs_1.of({})), operators_1.shareReplay());
}
function getOptionFromNpmRc(option) {
    return _readNpmRc().pipe(operators_1.map(options => options[option]));
}
function getOptionFromNpmCli(option) {
    return new rxjs_1.Observable(subject => {
        child_process_1.exec(`npm get ${option}`, (error, data) => {
            if (error) {
                throw error;
            }
            else {
                data = data.trim();
                if (!data || data === 'undefined' || data === 'null') {
                    subject.next();
                }
                else {
                    subject.next(data);
                }
            }
            subject.complete();
        });
    }).pipe(operators_1.catchError(() => rxjs_1.of(undefined)), operators_1.shareReplay());
}
function getNpmConfigOption(option, scope, tryWithoutScope) {
    if (scope && tryWithoutScope) {
        return rxjs_1.concat(getNpmConfigOption(option, scope), getNpmConfigOption(option)).pipe(operators_1.filter(result => !!result), operators_1.defaultIfEmpty(), operators_1.first());
    }
    const fullOption = `${scope ? scope + ':' : ''}${option}`;
    let value = npmConfigOptionCache.get(fullOption);
    if (value) {
        return value;
    }
    value = option.startsWith('_')
        ? getOptionFromNpmRc(fullOption)
        : getOptionFromNpmCli(fullOption);
    npmConfigOptionCache.set(fullOption, value);
    return value;
}
function getNpmClientSslOptions(strictSsl, cafile) {
    const sslOptions = {};
    if (strictSsl === 'false') {
        sslOptions.strict = false;
    }
    else if (strictSsl === 'true') {
        sslOptions.strict = true;
    }
    if (cafile) {
        sslOptions.ca = fs_1.readFileSync(cafile);
    }
    return sslOptions;
}
/**
 * Get the NPM repository's package.json for a package. This is p
 * @param {string} packageName The package name to fetch.
 * @param {string} registryUrl The NPM Registry URL to use.
 * @param {LoggerApi} logger A logger instance to log debug information.
 * @returns An observable that will put the pacakge.json content.
 * @private
 */
function getNpmPackageJson(packageName, registryUrl, logger) {
    const scope = packageName.startsWith('@') ? packageName.split('/')[0] : undefined;
    return (registryUrl ? rxjs_1.of(registryUrl) : getNpmConfigOption('registry', scope, true)).pipe(operators_1.map(partialUrl => {
        if (!partialUrl) {
            partialUrl = 'https://registry.npmjs.org/';
        }
        const partial = url.parse(partialUrl);
        let fullUrl = new url.URL(`http://${partial.host}/${packageName.replace(/\//g, '%2F')}`);
        try {
            const registry = new url.URL(partialUrl);
            registry.pathname = (registry.pathname || '')
                .replace(/\/?$/, '/' + packageName.replace(/\//g, '%2F'));
            fullUrl = new url.URL(url.format(registry));
        }
        catch (_a) { }
        logger.debug(`Getting package.json from '${packageName}' (url: ${JSON.stringify(fullUrl)})...`);
        return fullUrl;
    }), operators_1.concatMap(fullUrl => {
        let maybeRequest = npmPackageJsonCache.get(fullUrl.toString());
        if (maybeRequest) {
            return maybeRequest;
        }
        const registryKey = `//${fullUrl.host}/`;
        return rxjs_1.concat(getNpmConfigOption('proxy'), getNpmConfigOption('https-proxy'), getNpmConfigOption('strict-ssl'), getNpmConfigOption('cafile'), getNpmConfigOption('_auth'), getNpmConfigOption('_authToken', registryKey), getNpmConfigOption('username', registryKey, true), getNpmConfigOption('password', registryKey, true), getNpmConfigOption('email', registryKey, true), getNpmConfigOption('always-auth', registryKey, true)).pipe(operators_1.toArray(), operators_1.concatMap(options => {
            const [http, https, strictSsl, cafile, token, authToken, username, password, email, alwaysAuth,] = options;
            const subject = new rxjs_1.ReplaySubject(1);
            const sslOptions = getNpmClientSslOptions(strictSsl, cafile);
            const auth = {};
            if (alwaysAuth !== undefined) {
                auth.alwaysAuth = alwaysAuth === 'false' ? false : !!alwaysAuth;
            }
            if (email) {
                auth.email = email;
            }
            if (authToken) {
                auth.token = authToken;
            }
            else if (token) {
                try {
                    // attempt to parse username and password from base64 token
                    // to enable Artifactory / Nexus-like repositories support
                    const parsedToken = Buffer.from(token, 'base64').toString('ascii');
                    const [extractedUsername, extractedPassword] = parsedToken.split(':');
                    if (extractedUsername && extractedPassword) {
                        auth.username = extractedUsername;
                        auth.password = extractedPassword;
                    }
                    else {
                        throw new Error('Unable to extract username and password from _auth token');
                    }
                }
                catch (ex) {
                    auth.token = token;
                }
            }
            else if (username) {
                auth.username = username;
                auth.password = password;
            }
            const client = new RegistryClient({
                proxy: { http, https },
                ssl: sslOptions,
            });
            client.log.level = 'silent';
            const params = {
                timeout: 30000,
                auth,
            };
            client.get(fullUrl.toString(), params, (error, data) => {
                if (error) {
                    subject.error(error);
                }
                subject.next(data);
                subject.complete();
            });
            maybeRequest = subject.asObservable();
            npmPackageJsonCache.set(fullUrl.toString(), maybeRequest);
            return maybeRequest;
        }));
    }));
}
exports.getNpmPackageJson = getNpmPackageJson;
```

After that the `ng update` should start working.

**Comment by dkabul** [comment]
Still having issue with @angular/cli 6.2.2

**Comment by dherges** [comment]
```bash
$ ng update --registry <url-to-private-registry>
```

Doesn#t that work?

**Comment by mrmaxsteel** [comment]
> ```shell
> $ ng update --registry <url-to-private-registry>
> ```
> 
> Doesn#t that work?

Not when private registry requires credentials.

**Comment by smnbbrv** [comment]
@mrmaxsteel could you please check the following:

1. where is located your npmrc? Project-level or home directory?
2. does your password contain symbol ':'? If yes, try the updated script at https://github.com/angular/angular-cli/issues/10624#issuecomment-421564909
3. could you remove the spaces around '=' (it looks like current script does not work with spaces yet)? So it should look like this:

```
registry=<url to artifactory>/api/npm/<repo>
_auth=xxxx
always-auth=true
email=xxxx
```

Thanks for collaborating! 👍

**Comment by mrmaxsteel** [comment]
@smnbbrv yes! It works without whitespace between `_auth = token` as mentioned here: https://github.com/angular/angular-cli/pull/12284#issuecomment-422011350

It would be great if you could ensure support for whitespaces before your PR is merged, as officially the .npmrc should support whitespaces (https://docs.npmjs.com/files/npmrc)

**Comment by smnbbrv** [comment]
done and updated the script at #10624 (comment)

**Comment by mrmaxsteel** [comment]
Tested updated script from comment above, works 👍Nice one @smnbbrv! 

@hansl any idea when this will be released?

**Comment by arm1n** [comment]
I'm experiencing the same with `Angular CLI 6.2.2` and private bitbucket repo like `git+ssh://git@bitbucket.org:<org>/<package>`.

**Comment by arm1n** [comment]
@smnbbrv No I didn't - I'm expecting that it work without patching code from node_modules ;) But I will do it so that you know if it's working for that case too - is your patch supposed to land in one of the  next versions?

**Comment by smnbbrv** [comment]
@arm1n I am outside of Angular release cycle scope, but if it works with your case too then it is more likely to push it further. If it does not, could you post your npmrc (of course with masked credentials)

**Comment by SV-efi** [comment]
I cannot have ng serve running without choking on file changes. 
I have to update to a specific version to get the fix. And I cannot "update" using a simple command. Because I happened to use font-awesome. So I cannot develop, I cannot upgrade. Do you guys see the problem ?

**Comment by logic01** [comment]
I was really hoping this was resolved before v7.0 dropped. Now I am once again butting against this issue.

**Comment by Ionaru** [comment]
The fix by @smnbbrv did not work for me, I still got `401 Unauthorized` when running `ng update @angular/cli @angular/core`. 

My .npmrc is really simple:
```
@fortawesome:registry=https://npm.fontawesome.com/
//npm.fontawesome.com/:_authToken=SOME_TOKEN_HERE
```

**Comment by Ionaru** [comment]
@SV-efi You can probably just change the `@angular/...` packages to a newer `6.X.X` version and run `npm install` to fix your issue.

**Comment by evanjmg** [comment]
+1, can't we just read the .npmrc on install or have an alternative flag that skips the npm check and simply updates angular?

**Comment by PMoransais** [comment]
@lonaru

Angular cli STILL does not work with Font-Awesome pro npm authentication
As for @lonaru the @smnbbrv fix does not work and error 401 Unauthorized is thrown wen updating Cli to version 7
I have no .npmrc locally, only a global one.

I spent a lot of time updating to 6.1 because of FontAwsome authentication and using Font-Awesome is again a problem. Can this be dealt with once and for all ?

**Comment by SV-efi** [comment]
Works for me ! Thanks @smnbbrv !

**Comment by smnbbrv** [comment]
Super! If this works for the rest of the guys, I guess the issue is pretty much solved.

Could someone of fontawesome users post a request on their repo (or whatever contact they provide) to include this setting into documentation? This is a (very) common configuration item for the private repos and, well, then I don't understand why it's not there yet.

**Comment by trojanc** [comment]
It might work for font-awesome (i wouldn't know i dont use it), but it still doesn't work for a npm **proxy** that **has authentication**. E.g when you have a nexus server keeping a "local" (to the office network cache) of the default npm repository - there where all the public stuff like angular gets downloaded from by default (explained in my previous https://github.com/angular/angular-cli/issues/10624#issuecomment-421544428) 

Steps to reproduce:
1) Login to your private npm proxy `npm login --registry=https://nexus.myorg.com/repository/npm"
2) Now try to upgrade your existing project `ng upgrade`
2.1) Even try `ng upgrade --registry=https://nexus.myorg.com/repository/npm"

After waiting quite a while it errors with "This request requires auth credentials. Run `npm login` and repeat the request." Looking at my nexus logs I can see it never even attempted to download from the registry.

Because `ng update` doesn't have a `--verbose` options, I can't really tell where it is trying to download from....

**Comment by smnbbrv** [comment]
@trojanc please post your `.npmrc`

**Comment by trojanc** [comment]
> @trojanc please post your `.npmrc`

```
//nexus.mydomain.co.za/repository/npm/:_authToken=NpmToken.xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx
always-auth=true
```

**Comment by trojanc** [comment]
I think I found it...very much hidden I'd say. For me the issue was the LOCATION of the `.npmrc' file. Angular does not search for the file in the default location on windows. 
Normally the file is in `C:\Users\username\.npmrc` instead Angular searches for the file in `C:\Users\username\AppData\Local\.npmrc`.
Copying the file to there solved my issue. Now just to remember that every change I make to the normal location of npmrc - like when using npm cli to add repos - to copy that file to where Angular is looking for it

**Comment by mixer2** [comment]
@smnbbrv `ng update` still doesn't work for me. the .npmrc is in the same directory as the angular.json and contains `always-auth=true` and an artifactory url for `registry`. there is also the `~/.npmrc` which contains the `_authToken`. i'm using yarn as `packageManager` but however also removed the `packageManager` config from `angular.json` to use npm.
installation and updating of dependencies via npm and yarn works fine (so auth data should be correct).
it takes some time and outputs then `This request requires auth credentials. Run npm login and repeat the request.`. Angular CLI version 7.0.2, yarn version v1.10.1.
any suggestions?

**Comment by smnbbrv** [comment]
@mixer2 this is not yet implemented case. Try to put _authToken into the registry's .npmrc or use only global .npmrc, should work for one or another but not both

**Comment by mixer2** [comment]
@smnbbrv thx for the quick reply. i tried to add the auth data to the .npmrc of the project and also to remove that one and add always-auth and registry to the ~/.npmrc. unfortunately, both combinations didn't work -.-
any other ideas?

**Comment by GeorgeKnap** [comment]
pretty much same experience here while running `ng update`:
```
Your global Angular CLI version (7.0.3) is greater than your local
version (6.1.1). The local Angular CLI version is used.

To disable this warning use "ng config -g cli.warnings.versionMismatch false".
Unexpected token ﻿ in JSON at position 0
﻿{"$id":"1","innerException":null,"message":"TF400813: Resource not available for anonymous access. Client authentication required.","typeName":"Microsoft.TeamFoundation.Framework.Server.UnauthorizedRequestException, Microsoft.TeamFoundation.Framework.Server","typeKey":"UnauthorizedRequestException","errorCode":0,"eventId":3000}
```

`always-auth=true` does not fix it
moving private `.npmrc` to `C:\Users\username\AppData\Local\.npmrc` does not work either

**Comment by ThYpHo0n** [comment]
Please double check your .npmrc especially when you use scoped repositories with scoped authTokens.
E.g. I had the following in my .npmrc
`@myscope:registry=https://private.artifactory.instance/api/npm/private-npm/
//artifacts.innogames.de/api/npm/innogames-npm/:_authToken=myAuthToken`
The problem is that https://github.com/angular/angular-cli/blob/master/packages/schematics/update/update/npm.ts#L73 is searching for the exact match of the scraped string "//private.artifactory.instance/" which doesn't match.

To the Angular devs: This could be solved by adding a fuzzy matching into the getOptionFromNpmRc function but could then lead to other side effects. Nevertheless this should be covered in a better way (by adding more examples) to the documentation.

**Comment by mixer2** [comment]
@ThYpHo0n thanks! that was exactly the issue i had... it's a bit unfortunate that they share the .npmrc config with npm/yarn but interpret it different in multiple ways. especially the `run npm login` hint in the error is misleading, because it will create the _authToken in antoher file and with another registry scope than `ng update` would expect.
so for now the imho most reliable way to go is to manually add the _authToken with manually modified scope to the projects .npmrc, do the updates and then remove the token again to not accidentally push it.

**Comment by clydin** [comment]
The team is aware of the issue and there is a pending PR (https://github.com/angular/angular-cli/pull/12526) which will vastly improve the current situation.  Unfortunately, it is currently blocked on an upstream issue (https://github.com/zkat/pacote/issues/163).

**Comment by kroeder** [comment]
I tried `7.1.0-beta.0` and don't get "unauthenticated" anymore but it still fails.

Am I too soon? :) I just saw there were changes in the latest release
If not: Go on reading

Instead `ng update @angular/core` now throws a 404 with 
```
404 Not Found - GET https://registry.npmjs.org/@myScope%2fmyPackage - Not found
```

The thing is, this is a package stored in our private npm registry and should not be fetched from npmjs.org

I tried to
```
ng update @angular/core --registry https://registry.my-private-registry.com/repository/next.npm.group
```
But it fails with `Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"`

I tried to remove my scope declaration for our registry
```
@myScope:registry=__REGISTRY_URL__
``` 
didn't work 

Tried `always-auth=true` => Didn't work

**Comment by rafa-as** [comment]
Same here! All solutions didn't work for me

**Comment by daniel-seitz** [comment]
My workaround was to only migrate (which worked) and then upgrade the packages manually.

ng update @angular/cli --migrateOnly
(maybe you need to move the entries inside the home/.npmrc or .yarnrc into the project's corresponding file)

After that went ok, I i,e, used yarn update @angular/cli@6.0.7
(working with yarn here)

Unfortunately I can't really find what else ng update does and hope that these two steps are sufficient.

**Comment by joshvito** [comment]
> I think I found it...very much hidden I'd say. For me the issue was the LOCATION of the `.npmrc` file. Angular does not search for the file in the default location on windows.
> Normally the file is in `C:\Users\username\.npmrc` instead Angular searches for the file in `C:\Users\username\AppData\Local\.npmrc`.
> Copying the file to there solved my issue. Now just to remember that every change I make to the normal location of npmrc - like when using npm cli to add repos - to copy that file to where Angular is looking for it

I tried a symlink in the AppData/Local dir pointing at the User directory, and it didn't change the error.

**Comment by mswilson4040** [comment]
Is there a timeline for a fix? I'm not in a hurry to upgrade to 7 so I'd rather wait for the right fix instead of playing around with the workarounds.

**Comment by clydin** [comment]
In additional to the previous PR mentioned above (which is currently available in the 7.1.0.beta.1).  Further improvements will be present in the next beta via https://github.com/angular/angular-cli/pull/12871.  This PR provides improved discovery of `.nmprc` files.

**Comment by kroeder** [comment]
fyi

Tested `7.1.0-rc.0` but still get `404 Not Found - GET https://registry.npmjs.org/@myScope%2fmyPackage - Not found` where the URL should point to a private registry

```
Angular CLI: 7.1.0-rc.0
Node: 10.13.0
OS: win32 x64
Angular: 7.0.1
... animations, cdk, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.10.3
@angular-devkit/build-angular      0.10.3
@angular-devkit/build-ng-packagr   0.10.3
@angular-devkit/build-optimizer    0.10.3
@angular-devkit/build-webpack      0.10.3
@angular-devkit/core               7.0.3
@angular-devkit/schematics         7.1.0-rc.0
@angular/cli                       7.1.0-rc.0
@angular/flex-layout               7.0.0-beta.19
@ngtools/json-schema               1.1.0
@ngtools/webpack                   7.0.3
@schematics/angular                7.1.0-rc.0
@schematics/update                 0.11.0-rc.0
ng-packagr                         4.4.0
rxjs                               6.3.3
typescript                         3.1.3
webpack                            4.23.1
```

**Comment by clydin** [comment]
@kroeder Can you provide the path to the `.npmrc` file that contains the relevant settings?

**Comment by kroeder** [comment]
Hey, sorry for the late response

**Default directory**
`C:\Users\%username%\.npmrc` 

**Contains**
```
registry=https://%registry%/repository/%npm-group-name%.npm.group/
//%registry%/repository/%npm-group-name%.npm.group/:_authToken=%auth-token%
//%registry%/repository/%npm-group-name%.npm.releases/:_authToken=%auth-token%
```
*%...% are not part of the content, I just replaced the actual values with %-placeholders*

**We use:** Nexus https://blog.sonatype.com/using-nexus-3-as-your-repository-part-2-npm-packages

**Comment by clydin** [comment]
@kroeder if you have some time,  could you put that `.npmrc` file in your project's directory and try running update?

@mswilson4040 are you on windows as well?

**Comment by mswilson4040** [comment]
@clydin I am on a mac.

**Comment by clydin** [comment]
@mswilson4040 did you receive the same result when trying the 7.1 RC?  Also could you provide the path to the relevant `.npmrc` as well?

**Comment by mswilson4040** [comment]
I haven't tried 7.1 yet. I eventually just removed the npmrc and all my modules, ran ng update and then put everything back in. For the npmrc file path, I have two locations. One location in my home directory  (which has my auth token) and then another one at my project directory root (registering the priate repo).

- .npmrc path in home dir: (/Users/<username>)
- .npmrc in project dir: (/Users/<username>/dev/<project name>/wwwroot/clientapp)

**Comment by danielleroux** [comment]
Still not working.
Tested with 7.0.6 and 7.1.0-rc.0

After i installed a private package from a private repository the `ng update` command failed always with:
`404 Not Found: @my-private-scrope/my-private-package`

**Comment by kroeder** [comment]
@daniel have you tried https://github.com/angular/angular-cli/issues/10624#issuecomment-439943098

Still had no time to test this..

**Comment by danielleroux** [comment]
@kroeder yes, but i get the same errror.

OS: macOS

**Comment by kroeder** [comment]
Cannot confirm, what exactly have you done to get it work?
I tested
**Global** `7.1.0-rc.0` / **Local** `~7.0.0` => "Unauthorized"
**Global** `7.1.0-rc.0` / **Local** `^6.0.0` => "Unauthorized"
**Global** `7.1.0-rc.0` / **Local** `7.1.0-rc.0` => "404 Not Found: @myprivatescope/my-private-package"
**Global** `7.1.0-rc.0` / **Local** `7.1.0-rc.0` / `.npmrc` in project-root => "404 Not Found: @myprivatescope/my-private-package"

**Comment by clydin** [comment]
Environment variable replacement is not currently supported (i.e., `${NPM_TOKEN_FONTAWESOME}` will stay as is).  Support is currently in development.

**Comment by catull** [comment]
@clydin You are spot-on, as usual.

If I add the token verbatim, `ng upgrade` actually works.

As I do depend on an internal framework to first upgrade to Angular 7 - it is based on 6.1 -, I am not in a hurry to upgrade the code base to 7 yet.

Thanks for clarifying it.

**Comment by kroeder** [comment]
🤔 Is it windows or why doesn't it work with my project?

**Comment by catull** [comment]
@kroeder 

Your current C:\Users\%username%\.npmrc is

> registry=https://%registry%/repository/%npm-group-name%.npm.group/
//%registry%/repository/%npm-group-name%.npm.group/:_authToken=%auth-token%
//%registry%/repository/%npm-group-name%.npm.releases/:_authToken=%auth-token%

Have you tried

```ini
registry=https://%registry%/repository/
//%registry%/repository/:_authToken=%auth-token%
```

Since we do not know what `%npm-group-name%` actually means/looks like.

Is is comparable to:

```ini
@fortawesome:registry=https://npm.fontawesome.com/
//npm.fontawesome.com/:_authToken=XXXXXXX
```

?

**Comment by kroeder** [comment]
You might want to check http://codeheaven.io/using-nexus-3-as-your-repository-part-2-npm-packages/

I have no deep knowledge, I just use it. Anyway: Using ```registry=https://%registry%/``` and then `ng update @angular/core` prints an HTML-Document to the console because this is the URL to the nexus web interface and the actual repo is in a sub-directory

**Comment by clydin** [comment]
@kroeder does your `.npmrc` literally have the `%`'s or are those only for comment replacement purposes?

**Comment by kroeder** [comment]
Just placeholders because I don't want to share the exact URL :) have no vars

**Comment by catull** [comment]
Hmm.

I have globally and locally @angular/cli 7.1.0.
Our project uses '@angular/*' "^6.1.0".

Did a clean `yarn install` - there was no yarn.lock or node_modules prior to the install.

Then I wanted to perform `ng update '@angular/cli'`.

I got this red line as result:

```log
401 Unauthorized - GET https://npm.fontawesome.com/@fortawesome%2fpro-regular-svg-icons
```

Same result with  `ng update '@angular/core'`.

As a reminder, our local .npmrc is (we do not use .npmrc in HOME directory):

```ini
@XXXXXXXXXX:registry=https://artifactory.YYYYYYYYYY:443/artifactory/api/npm/npm-virtual/

@fortawesome:registry=https://npm.fontawesome.com/
//npm.fontawesome.com/:_authToken=${NPM_TOKEN_FONTAWESOME}

PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true

depth=0
loglevel=warn
package-lock=false
progress=false
```

It only works if I replace ${NPM_TOKEN_FONTAWESOME} with the real token string.

I take it @clydin's enhancement to support ENV variables in .npmrc has not landed yet.

**Comment by clydin** [comment]
@catull The environment variable support will be in 7.1.1.

**Comment by catull** [comment]
Excellent, thanks!

**Comment by ManishJainGitHubUser** [comment]
+1 same issue with 7.1.0 - @microsoft fix it

**Comment by denisoby** [comment]
The problem for angular-cli 7.1.0 is in angular-cli/packages/schematics/update/update/npm.ts

```
const resultPromise = pacote.packument(
    packageName,
    {
      'full-metadata': true,
      ...npmrc,
      registry: options && options.registryUrl,
    },
  );
```

This line contains error: 
```registry: options && options.registryUrl```

If you pass any cli options, but don't pass registryUrl inside this options - this line will generate "registry: undefined" that always overrides ".npmrc" value.

And pacote when see this - uses default registry.npmjs.org

The fix is simple - check that options.registryUrl has some value, before passing it to pacote.
I can try to make a fix, and also some tests for it :)

**Comment by catull** [comment]
@denisoby Don't bother.

@hansl committed a fix with https://github.com/angular/angular-cli/commit/6fce79355fb6305f3db2e627fdf345b9ac1f173b
@clydin enhanced it with https://github.com/angular/angular-cli/pull/13106/commits/e85da635cd70ddcf1eddee430a7d740eaa3bc53e

Just wait for version @angular/cli@7.1.1

**Comment by DaSchTour** [comment]
With 7.1.1 I get 404 errors again. Scoped packages seams not to be recognized correctly.

**Comment by cluen** [comment]
Is there an article describing what to do know?
Im using Windows and don't know how to get it working: 
- Use .npmrc in local and/or HOME ?
- Do I have to use environment variables for auth-keys?

Deleted my local .nmprc to update version of @angular/cli to 7.1.1. 
Restored the local .npmrc and trying to start ng update

If I then try to start ng update getting also the 404 with scoped package names:
like: `/@angular-devkit%2fbuild-angular`

**Comment by alan-agius4** [comment]
This should have been fixed by https://github.com/angular/angular-cli/pull/13025

**Comment by Albyzai** [comment]
@alan-agius4 I am running 7.1.2 and I still have this problem with a private repo.

**Comment by alan-agius4** [comment]
@Albyzai can you kindly do `ng --version` you should have `@schematics/update` version `0.11.2`

**Comment by alan-agius4** [comment]
@Albyzai can you kindly try to use the `next` version? Note that you should also update @angular/devkit packages

**Comment by Albyzai** [comment]
@alan-agius4 

This is my output:

```
@angular-devkit/architect         0.10.7
@angular-devkit/build-angular     0.10.7
@angular-devkit/build-optimizer   0.10.7
@angular-devkit/build-webpack     0.10.7
@angular-devkit/core              7.0.7
@angular-devkit/schematics        7.1.2
@angular/cdk                      7.1.1
@ngtools/webpack                  7.0.7
@schematics/angular               7.1.2
@schematics/update                0.11.2
rxjs                              6.3.3
typescript                        3.1.1
webpack                           4.19.1
```

**Comment by alan-agius4** [comment]
@Albyzai, can you try to use  `@angular/cli@next` and see if the problem persists?

**Comment by Albyzai** [comment]
@alan-agius4 

```
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 7.2.0-beta.1
Node: 8.12.0
OS: win32 x64
Angular: 7.1.2
... animations, common, compiler, compiler-cli, core, forms
... http, platform-browser, platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.10.7
@angular-devkit/build-angular     0.10.7
@angular-devkit/build-optimizer   0.10.7
@angular-devkit/build-webpack     0.10.7
@angular-devkit/core              7.0.7
@angular-devkit/schematics        7.2.0-beta.1
@angular/cdk                      7.1.1
@angular/cli                      7.2.0-beta.1
@ngtools/webpack                  7.0.7
@schematics/angular               7.2.0-beta.1
@schematics/update                0.12.0-beta.1
rxjs                              6.3.3
typescript                        3.1.1
webpack                           4.19.1
```

Problem still persists.

**Comment by daniel-seitz** [comment]
btw. is this going to be released into ^6.x.x?

There are quite some repos still out there on versions before 6 and the angular-team behind https://update.angular.io doesn't recommend to move across multiple major versions.

Hence it would be great to get it into 6 as well.

**Comment by catull** [comment]
It works for me, since 7.1.1.

@Albyzai Care to show us your .npmrc, mask out the spicy bits.

**Comment by Albyzai** [comment]
@catull 

Will do on Monday, got it on my work PC.

**Comment by daniel-seitz** [comment]
With @next We get the message ´request to https://registry.npmjs.org/@types%2fjasmine failed, reason: connect ETIMEDOUT´

It seems that ´ng update @angular/cli´ ignores the proxy since it should go to the proxy or https-proxy instead of npmjs.org.

Should the command be ´ng update @angular/cli@next´? (Results in same error message)

**Comment by catull** [comment]
@alex88 @daniel-seitz
You two are aware that we cannot help you if you do not describe your situation adequately enough.

For starters, post your .npmrc.
Mention if you have an .npmrc in your home directory and/or in your project root folder.

Mask out the portions you DO NOT want us to know, like internal host names or credentials etc.
Just give enough pieces of information.

**Comment by daniel-seitz** [comment]
@catull thanks for clarifying. I thought we're on the same page already.

1. we use yarn.
2. local .npmrc:
   `registry=[some url, masked out]
   always-auth=true`
3. yarn lock contains packages where I see some as resolved with [some url, masked out]
   also there are packages with `https://registry.yarnpkg.com/some-package-and-so-on`
4. global .yarnrc (in C:\Users\MyUserName):
`# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1
lastUpdateCheck 1543945019859
`
5. global .npmrc (in C:\Users\MyUserName):
`//some_path_with_some_token
proxy= http://our_proxy_with_port/
https-proxy= http://our_proxy_with_port
`
6. trying to follow commands from update.angular.io
7. everything works but not `ng update @angular/cli`, message comes up after a while (1 minute or so):
`request to https://registry.npmjs.org/@types%2fjasmine failed, reason: connect ETIMEDOUT`
seems to be on a random package, at another try I got:
`request to https://registry.npmjs.org/codelyzer failed, reason: connect ETIMEDOUT [some ip address and port]`
8. currently still running angular 5.1.1
9. tried as described with @next option in `yarn global add @angular/cli@next` and `yarn add @angular/cli@next
`
10. ng --version:
`
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 7.2.0-beta.1
Node: 8.12.0
OS: win32 x64
Angular: 5.1.1
... animations, common, compiler, compiler-cli, core, forms
... http, platform-browser, platform-browser-dynamic
... platform-server, router

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.12.0-beta.1
@angular-devkit/core         7.2.0-beta.1
@angular-devkit/schematics   7.2.0-beta.1
@angular/cli                 7.2.0-beta.1
@angular/language-service    4.4.6
@schematics/angular          7.2.0-beta.1
@schematics/update           0.12.0-beta.1
rxjs                         5.5.3
typescript                   2.4.2`

**Comment by alex88** [comment]
@catull well `.npmrc` isn't used when having a git dependency. it's just a line like this:

```
"my-package": "git+ssh://git@gitlab.com:my-package.git#master"
```

in the `package.json`. So package repo is involved.

**Comment by catull** [comment]
@alex88

I can reproduce your behaviour, I think.

I have no `.npmrc` in my home directory - I renamed it to `.npmrc.old`,

I have created a brand-new angular app "foo", there is no `.npmrc` in that project folder.
In that folder, I can perform `ng upgrade`, no error is reported.

Then I added this dependency:
```json
    "@streamlabs/obs-studio-node": "git+ssh://git@gitlab.com:computerquip/obs-studio-node#staging",
```

Now, when I perform `ng update`, the result is ....

```log
➜  foo git:(master) ✗ ng update
404 Not Found - GET https://registry.npmjs.org/@streamlabs%2fobs-studio-node - Not found
```

This is pretty much what you experience, with a different dependency.

`ng` fails in calculating the dependency graph for the newly introduced "@streamlabs/obs-studio-node".

I also tried with `ng update --registry=https://registry.npmjs.org` or `ng update --registry https://registry.npmjs.org`.

The result is the same as with `ng update`.

**Comment by catull** [comment]
@daniel-seitz 

Your situation is that you do not get a proxy'd connection, it times out.
Are you certain you have the correct port configuration, that it is not blocked by a firewall ?

Perhaps that proxy may only be used for "normal" HTTP ("browser") requests ...

To address your point 7, each package is equally affected, there is no guarantee that the packages are processed top to bottom.

Imagine `ng` fires off 20 requests asynchronously, the first one that times out is reported and `ng` errors out.
Once it is the 7th, next time is is the 13th, and then it is the 4th alphabetically.
Which one of those 20 it is, is entirely "random".

My best guess is: the proxy configuration is not correct.

Are you sure you need a proxy configuration ?

Unless it is a typo, I find it suspicious that you use a "http" proxy value for "https" for starters.

Can you exclude the fact that neither HTTP_PROXY nor HTTPS_PROXY *environment* variables are set.
Neither http_proxy / https_proxy *shell* variables are set, either.

**Comment by alex88** [comment]
@catull probably it won't handle any other git repository right? Like gitlab/github/bitbucket/aws codecommit and many others...

**Comment by catull** [comment]
@alex88 Your repo URL ...

> "my-package": "git+ssh://git@gitlab.com:my-package.git#master"

should it not be

"my-package": "git+ssh://git@gitlab.com:>>>>USERNAME/<<<<<<my-package.git#master"

Compare this to my example: 

  "@streamlabs/obs-studio-node": "git+ssh://git@gitlab.com:>>>>>>computerquip/<<<<<<obs-studio-node#staging",

**Comment by alex88** [comment]
@catull yes it's actually username/repo, I just wanted to remove also that part, sorry for the confusion

**Comment by catull** [comment]
No worries, just wanted to exclude another explanation ....

**Comment by catull** [comment]
@alex88 Have you considered using _scoped_ package naming [1] ?

See, this is a snippet of my .npmrc:

@XXXXXXXXXX:registry=https://artifactory.YYYYYYYYYY:443/artifactory/api/npm/npm-virtual/

```ini
@fortawesome:registry=https://npm.fontawesome.com/
//npm.fontawesome.com/:_authToken=${NPM_TOKEN_FONTAWESOME}
```

You could have

"my-package": "git+ssh://git@gitlab.com:@YOUR_SCOPE/my-package.git#master"

As long as the package is published with @YOUR_SCOPE - could be @YOUR_ORGANISATION.

It works for FontAwesome.
Their scope name @forteawesome - yeah, the scope if "FORTE awesome", not "font awesome".

Side note: You and I both know that "forte" in colloquial Italian means "awesome" in colloquial American English.

FontAwesome's repository is not the canonical from npmjs.org
Yet `npm`, `yarn` and most importantly `ng` all retrieve it from the right place.

Buona fortuna!

[1] https://docs.npmjs.com/misc/scope

**Comment by catull** [comment]
@daniel-seitz

Perhaps you are right.

Assuming the http proxy config works for, say, the initial `npm install`.
If that is the case, your issue is that `ng update` ignores the proxy configuration.
Then I apologise, my analysis above would be incorrect.

However, it is a different case from the original posting.

I suggest you open a new issue with exactly the the title "`ng update` ignores proxy config."
I do not have a setup which requires a proxy configuration, so I cannot assist you, sorry.

**Comment by alex88** [comment]
@catull haha for the side note 🤣
Sometimes I read it as awesome-awesome and sometimes if feels like a typo 😆

Anyway, that would involve having our own registry or paying for a private hosted on if I understand it correctly, I think that for now I'll just remove that dependency, upgrade to angular 7 and add it back 😄

Hate to do that but having our own means more maintenance/resources and the other option would be adding another cost. You probably also know how much Italian companies can be stingy 😄 

Btw, thanks for the option, if it's something we need to support in long term that's definitely an option to look for!

**Comment by catull** [comment]
@alex88

On the scoped naming, the cost you incur is operating an NPM registry.

See open source alternatives below, "verdaccio" [1] and "bitsrc.io" [2].
I believe there are docker images for both of them.
At one point, "sinopia" was actively developed, but it has not been updated for more than 3 years.

Imagine this.

Let's just assume for a moment I like a certain library/package.

Let's say it is called `banana`; the original `npm` package would be available at https://www.npmjs.com/package/banana.

So, one day I decide to considerably change it, but the original maintainer is
a) not around, the e-mail she is using is invalid by now
b) the maintainer is not interested in your changes
c) the package is not maintained any more, and maintainership cannot be transferred to you

Now in this case, `banana` was last updated 7 years ago.
But that is beside the point.

Given the fact that I like `banana`, I can create a fork of it.
I will NOT be able to publish it at https://www.npmjs.com/package/banana.

But I will be able to publish it say, as https://www.npmjs.com/package/@catull/banana.
Then it would be public.

But let's also assume my company operates an internal NPM repository at registry.my-company.net:8080/npm/.

Now I can publish it at https://registry.my-company.net:8080/npm/@SOME_SCOPE/banana

You have to publish it with `--scope @SOME_SCOPE`.
Otherwise you would publish it at https://registry.my-company.net:8080/npm/banana

But here is the beauty: the scope name can be specifically configured in `.npmrc`.
It is recognised by all tools equally: `npm`, `yarn` and `ng`.

You can see a good example with [bootstrap-vue](https://www.npmjs.com/search?q=bootstrap-vue):
- the original [bootstrap-vue](https://www.npmjs.com/package/bootstrap-vue)
- a fork [@aorinevo/bootstrap-vue](https://www.npmjs.com/package/@aorinevo/bootstrap-vue)

[1] [verdaccio](https://github.com/verdaccio/verdaccio)
[2] [bitsrc.io](https://hackernoon.com/private-npm-registry-simplified-with-bit-cc80408f1809)

**Comment by alex88** [comment]
@catull thanks a lot for the long explanation, I'll keep in mind, surely it's better than fetching from git directly 😄

**Comment by DaSchTour** [comment]
I really wonder why fontawesome pro works with 7.1.0 but fails with 7.1.1 and 7.1.2
It's a bit sad to see that newer version break again. I hope this can be finally fixed soon.

**Comment by Chklang** [comment]
I've same problem. "ng update --all" doesn't works with 7.1.2 (and 7.1.0 also) with private registry configured into local ".npmrc". @angular/cli while try to found package on registry.npmjs.org:

404 Not Found - GET https://registry.npmjs.org/mypackage - Not found

But into .npmrc i've :
registry=https://privaterepo.mycompagnie.fr/npm/

**Comment by aguacongas** [comment]
@Chklang did you try with `ng update --registry https://privaterepo.mycompagnie.fr/npm/` ?

**Comment by kpaxton** [comment]
@aguacongas that works for me!
However, I shouldn't have to do that if the registry is declared in my `.npmrc`.

**Comment by aguacongas** [comment]
@kpaxton You're lucky, that doesn't for me. But my private repo needs authentication.
And I agree, we shouldn't have to do that.

**Comment by aniruddhadas9** [comment]
I also not lucky. I have also authentication set for my private registry and I am getting below error 

`Unable to authenticate, need: Basic realm="Artifactory Realm"`

**Comment by aniruddhadas9** [comment]
> @Chklang did you try with `ng update --registry https://privaterepo.mycompagnie.fr/npm/` ?

Do you what what does `ng update --registry https://privaterepo.mycompagnie.fr/npm/` does so that for users having auth set in there private registry can directly do what this command does?

**Comment by joeson1** [comment]
In our npmrc we have options like strict-ssl and always-auth which are boolean values when being parsed. The update script in npm.ts tries to call 'replace' on the options and fails when the function does not exist (which it doesn't for bools):
`options[key] = options[key].replace(/\$\{([^\}]+)\}/, (_, name) => process.env[name] || '');`
We fixed this by checking if the function exists before trying to call it. PR is attached

Now, ng update works as expected also with our private Artifactory

**Comment by florian-kittel** [comment]
Hi,

I had the same problem, because we are using Artifactory in our company. We store private repositories in it. So I had the same error when trying to use ng update. 

But I can pass the parameter "-f" to force the update. I changed the names to give you an example what will happen with the force flag: 
```
ng update @angular/cli -f
    Package not installed: "-f". Skipping.
    404 Not Found - Get https://registry.npmjs.org/@MYCORPORATE/MYREPO1 - Not Found
    404 Not Found - Get https://registry.npmjs.org/@MYCORPORATE/MYREPO2 - Not Found
```

After that all was installed and update. I did this with the Angular and CDK package as well. We are updated the internal packages for our own.

I hope this will help.

**Comment by florian-kittel** [comment]
PS: We also use https://www.npmjs.com/package/npm-check-updates with ncu command to check updates for our packages. This works well with Artifactory. So we now using ng update for all Angular related packages and ncu for other. But be careful and check if all new packages are working with the Angular packages.

**Comment by catull** [comment]
@florian-kittel 
Both `npm` and `yarn` have had the `outdated` option, check it out, it may satisfy your requirements.
Works on public and private `npm` repos.

**Comment by DaSchTour** [comment]
Currently with `7.3.0` for me it works correctly again.

**Comment by clydin** [comment]
This should now be addressed in 7.3.0+.  If anyone is still encountering problems, please open a new issue detailing the issue.
